BJS2-85531

saserka opened the pull request:
- реализовал битву, где 2 война дерутся и их файт выводится на экран
- не реализовал: Имя передается извне(не могу понять, что это значит, может и правильно сделал, но не уверен)
- не реализовал цикл, что у противника не может быть меньше 0 (цикл сам сделал, но не знаю, как его вызвать)

saserka committed:
BJS2-85531
- реализовал битву, где 2 война дерутся и их файт выводится на экран
- не реализовал: Имя передается извне(не могу понять, что это значит, может и правильно сделал, но не уверен)
- не реализовал цикл, что у противника не может быть меньше 0 (цикл сам сделал, но не знаю, как его вызвать)

saserka committed:
BJS2-85531
- реализовал битву, где 2 война дерутся и их файт выводится на экран
- реализовал: Имя передается извне
- реализовал цикл, что у противника не может быть меньше 0

20cognosce commented on src/main/java/school/faang/bjs2_85531/Archer.java:
```diff
@@ -0,0 +1,21 @@
+package school.faang.bjs2_85531;
+
+public class Archer extends Character {
+    private static final int strength = 3;
+    private static final int agility = 10;
+    private static final int intelligence = 5;
```
вынести в отдельный класс с public static final константами и импортировать их в классы-герои

20cognosce replied:
константы пишутся в UPPER_SNAKE_CASE

20cognosce commented on src/main/java/school/faang/bjs2_85531/Archer.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public void attack(Character opponent) {
+        health = health - opponent.getStrength();
```
ты оппонента должен бить, а не себя

20cognosce commented on src/main/java/school/faang/bjs2_85531/Archer.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    @Override
+    public void attack(Character opponent) {
```
у тебя 90% этого метода дублируется в Warrior и Archer

вынеси все общее в базовый класс (в метод attack) и добавь только абстрактный метод getHeroAttackDamage() который возвращает число - урон текущего героя

этот урон используй в методе attack

20cognosce commented on src/main/java/school/faang/bjs2_85531/Warrior.java:
```diff
@@ -0,0 +1,22 @@
+package school.faang.bjs2_85531;
+
+public class Warrior extends Character {
+    private static final int strength = 9;
+    private static final int agility = 5;
+    private static final int intelligence = 3;
+
+    public Warrior(String name) {
+        super(name, strength, agility, intelligence);
+    }
+
+
+    @Override
+    public void attack(Character opponent) {
+        health = health - opponent.getAgility();
+        if (isKilling()) {
+            System.out.println(getName() + " повержен");
+            return;
+        }
+        System.out.println(getName() + " " + health + " ХП");
```
Аналогично

20cognosce commented on src/main/java/school/faang/bjs2_85531/Character.java:
```diff
@@ -0,0 +1,37 @@
+@Getter
+@Setter
+public abstract class Character {
+    private String name;
+    private int strength;
+    private int agility;
+    private int intelligence;
```
final

20cognosce commented on src/main/java/school/faang/bjs2_85531/Character.java:
```diff
@@ -0,0 +1,37 @@
+    private int strength;
+    private int agility;
+    private int intelligence;
+    protected int health = 100;
```
100 - константа

20cognosce commented on src/main/java/school/faang/bjs2_85531/Character.java:
```diff
@@ -0,0 +1,37 @@
+
+    public Character(String name) {
+        this.name = name;
+        this.strength = 5;
+        this.agility = 5;
+        this.intelligence = 5;
```
константы

20cognosce commented on src/main/java/school/faang/bjs2_85531/Character.java:
```diff
@@ -0,0 +1,37 @@
+    private int agility;
+    private int intelligence;
+    protected int health = 100;
+
+
+
```
что с пробелами, почему столько

20cognosce commented on src/main/java/school/faang/bjs2_85531/Character.java:
```diff
@@ -0,0 +1,37 @@
+
+    public abstract void attack(Character opponent);
+
+    public boolean isKilling() {
+        return health <= 0;
+    }
```
не должно быть здоровья меньше 0
если урона больше - выставляешь 0, но не в минус уносишь

20cognosce commented on src/main/java/school/faang/bjs2_85531/Main.java:
```diff
@@ -0,0 +1,17 @@
+            warrior.attack(archer);
+
+            archer.attack(warrior);
+            if (warrior.isKilling() || archer.isKilling()) {
```
что значит isKilling ? 
я перевожу это как "убивает"

boolean методы / переменные должны отвечать на вопрос да / нет

давай понятные имена своим методам и переменным

Commit:
saserka-BJS2-85531
- реализовал битву, где 2 война дерутся и их файт выводится на экран
- реализовал: Имя передается извне
- реализовал цикл, что у противника не может быть меньше 0

Commit:
saserka-BJS2-85531
- реализовал битву, где 2 война дерутся и их файт выводится на экран
- реализовал: Имя передается извне
- реализовал цикл, что у противника не может быть меньше 0

20cognosce commented on src/main/java/school/faang/bjs2_85531/Warrior.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @Override
+    public void attack(Character opponent) {
+        if (opponent.isAlive()) {
+            opponent.health = opponent.getHealth() - opponent.getHeroAttackDamage();
+            if (opponent.getHealth() <= 0) {
+                opponent.health = 0;
+                System.out.println(opponent.getName() + " повержен");
+                return;
+            }
+            System.out.println(opponent.getName() + " " + opponent.getHealth() + " ХП");
+        }
+    }
```
ну у тебя опять код дублируется, вынеси это в базовый класс Character, метод attack реализуй, можно не делать его абстрактным, достаточно только getHeroAttackDamage сделать абстрактным

20cognosce commented on src/main/java/school/faang/bjs2_85531/Warrior.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    public void attack(Character opponent) {
+        if (opponent.isAlive()) {
+            opponent.health = opponent.getHealth() - opponent.getHeroAttackDamage();
```
ну тут у тебя баг

ты оппоненту должен уменьшить здоровье на размер урона твоего героя, зачем ты у оппонента урон считываешь?

Далее, к переменным напрямую лучше не обращаться - opponent.health, это нарушает инкапсуляцию
Можешь здоровье сделать приватным и через сеттер его устанавливать

20cognosce commented on src/main/java/school/faang/bjs2_85531/Warrior.java:
```diff
@@ -0,0 +1,29 @@
+    public void attack(Character opponent) {
+        if (opponent.isAlive()) {
+            opponent.health = opponent.getHealth() - opponent.getHeroAttackDamage();
+            if (opponent.getHealth() <= 0) {
+                opponent.health = 0;
```
не должно такого быть!!!! сначала выполни проверки, затем выполняй один раз действие

почему ты два раза здоровье переприсваиваешь

20cognosce commented on src/main/java/school/faang/bjs2_85531/Warrior.java:
```diff
@@ -0,0 +1,29 @@
+                System.out.println(opponent.getName() + " повержен");
+                return;
+            }
+            System.out.println(opponent.getName() + " " + opponent.getHealth() + " ХП");
```
не складывай строки через +, используй String.format

20cognosce commented on src/main/java/school/faang/bjs2_85531/Warrior.java:
```diff
@@ -0,0 +1,29 @@
+        super(name, STRENGTH, AGILITY, INTELLIGENCE);
+    }
+
+    @Override
+    public int getHeroAttackDamage() {
+        return getStrength();
+    }
```
Это хорошо

20cognosce commented on src/main/java/school/faang/bjs2_85531/Warrior.java:
```diff
@@ -0,0 +1,29 @@
+package school.faang.bjs2_85531;
+
+public class Warrior extends Character {
+    public static final int STRENGTH = 5;
+    public static final int AGILITY = 5;
+    public static final int INTELLIGENCE = 5;
```
Все константы вынести в отдельный класс с константами, не размазывай их по 3-м классам, в одно место вынеси

20cognosce commented on src/main/java/school/faang/bjs2_85531/Archer.java:
```diff
@@ -0,0 +1,29 @@
+package school.faang.bjs2_85531;
+
+public class Archer extends Character {
```
Аналогично

20cognosce commented on src/main/java/school/faang/bjs2_85531/Character.java:
```diff
@@ -0,0 +1,39 @@
+@Getter
+@Setter
+public abstract class Character {
+    public static final int STRENGTH = 5;
+    public static final int AGILITY = 5;
+    public static final int INTELLIGENCE = 5;
+    public static final int HELTH = 100;
```
.

20cognosce commented on src/main/java/school/faang/bjs2_85531/Character.java:
```diff
@@ -0,0 +1,39 @@
+        this.intelligence = intelligence;
+    }
+
+    public abstract void attack(Character opponent);
```
ты можешь общую логику сюда вынести в метод attack

Commit:
saserka-BJS2-85531

20cognosce commented on src/main/java/school/faang/bjs2_85531/Character.java:
```diff
@@ -0,0 +1,52 @@
+@Setter
+public abstract class Character {
+    private int health = HEALTH;
+    private String name;
```
final

20cognosce commented on src/main/java/school/faang/bjs2_85531/Character.java:
```diff
@@ -0,0 +1,52 @@
+    public void attack(Character opponent) {
+        if (opponent.isAlive()) {
+            int damage = this.getHeroAttackDamage();
+            opponent.setHealth(opponent.getHealth() - damage);
+            if (opponent.getHealth() == 0) {
```
почему ровно 0 проверяешь, это некорректно
у тебя в setHealth можно -1000 поставить здоровье

нельзя устанавливать здоровье меньше 0, 0 - это минимум

Commit:
saserka-BJS2-85531
